// botbuilder/lib/eventFactory.js
"use strict";
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const botbuilder_core_1 = require("botbuilder-core");
const moment = require("moment-timezone");
const handoffEventNames_1 = require("./handoffEventNames");
/**
 * Contains utility methods for creating various event types.
 */
class EventFactory {
    /**
     * Create handoff initiation event.
     * @param context The context object for the turn.
     * @param handoffContext Agent hub-specific context.
     * @param transcript Transcript of the conversation.
     */
    static createHandoffInitiation(context, handoffContext, transcript) {
        if (!context) {
            throw new TypeError('EventFactory.createHandoffInitiation(): Missing context.');
        }
        const handoffEvent = this.createHandoffEvent(handoffEventNames_1.HandoffEventNames.InitiateHandoff, handoffContext, context.activity.conversation);
        handoffEvent.from = context.activity.from;
        handoffEvent.relatesTo = botbuilder_core_1.TurnContext.getConversationReference(context.activity);
        handoffEvent.replyToId = context.activity.id;
        handoffEvent.serviceUrl = context.activity.serviceUrl;
        handoffEvent.channelId = context.activity.channelId;
        if (transcript) {
            const attachment = {
                content: transcript,
                contentType: 'application/json',
                name: 'Transcript'
            };
            handoffEvent.attachments.push(attachment);
        }
        return handoffEvent;
    }
    /**
     * Create handoff status event.
     * @param conversation Conversation being handed over.
     * @param state State, possible values are: "accepted", "failed", "completed".
     * @param message Additional message for failed handoff.
     */
    static createHandoffStatus(conversation, state, message) {
        if (!conversation) {
            throw new TypeError('EventFactory.createHandoffStatus(): missing conversation.');
        }
        if (!state) {
            throw new TypeError('EventFactory.createHandoffStatus(): missing state.');
        }
        const value = { state, message };
        const handoffEvent = this.createHandoffEvent(handoffEventNames_1.HandoffEventNames.HandoffStatus, value, conversation);
        return handoffEvent;
    }
    static createHandoffEvent(name, value, conversation) {
        const handoffEvent = {};
        handoffEvent.name = name;
        handoffEvent.value = value;
        handoffEvent.id = uuid();
        handoffEvent.timestamp = new Date(Date.now());
        // The timestamp does not contain the local offset which is a known limitation of Date objects in JavaScript.
        // Therefore, the localTimezone is included in the handoffEvent.
        handoffEvent.localTimezone = moment.tz.guess();
        handoffEvent.conversation = conversation;
        handoffEvent.attachments = [];
        handoffEvent.entities = [];
        return handoffEvent;
    }
}
exports.EventFactory = EventFactory;
function uuid() {
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
        let r = Math.random() * 16 | 0, v = c == 'x' ? r : (r & 0x3 | 0x8);
        return v.toString(16);
    });
}
