// botframework-streaming/browser/utilities/createNodeServer.js
"use strict";
/**
 * @module botframework-streaming
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createNodeServer = function (callback) {
    if (callback && typeof callback !== 'function') {
        throw new TypeError("Invalid callback; callback parameter must be a function to create Node 'net' Server.");
    }
    try {
        var server = exports.getServerFactory()(callback);
        if (isNetServer(server)) {
            return server;
        }
    }
    catch (error) {
        throw error;
    }
};
exports.getServerFactory = function () {
    if (typeof require !== undefined) {
        return require('net').Server;
    }
    throw TypeError("require is undefined. Must be in a Node module to require 'net' dynamically in order to fetch Server factory.");
};
function isNetServer(o) {
    return (hasCloseMethod && hasListenMethod) ? true : false;
}
function hasCloseMethod(o) {
    return (o.close && typeof o.close === 'function') ? true : false;
}
function hasListenMethod(o) {
    return (o.listen && typeof o.listen === 'function') ? true : false;
}
